numpy/lib/function_base.py

```diff
@@ -1624,46 +1624,45 @@ def trim_zeros(
     """
     trim = trim.lower()
 
+    absolutes = np.abs(filt)
+    lengths = np.zeros((absolutes.ndim, 2), dtype=np.intp)
+
     if axis is None:
         # Apply iteratively to all axes
-        axis = range(filt.ndim)
-
+        axis = range(absolutes.ndim)
     # Normalize axes to 1D-array
     axis = np.asarray(axis, dtype=np.intp)
     if axis.ndim == 0:
         axis = np.asarray([axis], dtype=np.intp)
 
-    absolutes = np.abs(filt)
-    lengths = np.zeros((absolutes.ndim, 2), dtype=np.intp)
+    if atol > 0:
+        absolutes[absolutes <= atol] = 0
+    if rtol > 0:
+        absolutes[absolutes <= rtol * absolutes.max()] = 0
+
+    nonzero = np.nonzero(absolutes)
 
     for current_axis in axis:
         absolutes.take([], current_axis)  # Raises if axis is out of bounds
         if current_axis < 0:
             current_axis += absolutes.ndim
 
-        # Reduce to envelope along all axes except the selected one
-        reduced = np.moveaxis(absolutes, current_axis, -1)
-        for _ in range(absolutes.ndim - 1):
-            reduced = reduced.max(axis=0)
-        assert reduced.ndim == 1
-
-        if atol > 0:
-            reduced[reduced <= atol] = 0
-        if rtol > 0:
-            reduced[reduced <= rtol * reduced.max()] = 0
-
-        # Find start and stop indices for current dimension
-        start, stop = np.nonzero(reduced)[0][[0, -1]]
-        stop += 1
+        if nonzero[current_axis].size > 0:
+            start = nonzero[current_axis].min()
+            stop = nonzero[current_axis].max()
+            stop += 1
+        else:
+            # In case the input is all-zero, slice only in front
+            start = stop = absolutes.shape[current_axis]
+            if "f" not in trim:
+                # except when only the backside is to be sliced
+                stop = 0
```

Comment:
I don't really understand what you mean here - are you saying to slice _leaving_ only the front, or slice _removing_ the front?

When asked to slice both ends of an all zero array, I think I'd expect either:
* Only the back to be sliced
* The front to be sliced if `'fb'` is passed, and the back to be sliced if `'bf'` is passed

Reply:
Good call. I prefer the latter behavior and have modified the code accordingly. 

The old function was really liberal with what it accepted as input for the `trim` parameter (e.g. only checking `"f" in trim`). I think restricting what is allowed to be passed in would make that parameter easier to understand and document.